src/porepy/viz/data_saving_model_mixin.py

```diff
@@ -315,3 +289,172 @@ def load_data_from_pvd(
+
+    def initialize_data_saving(self):
+        """Initialize iteration exporter."""
+        super().initialize_data_saving()
+        # Having a separate exporter for iterations avoids distinguishing between
+        # iterations and time steps in the regular exporter's history.
+        self.iteration_exporter = pp.Exporter(
+            self.mdg,
+            file_name=self.params["file_name"],
+            folder_name=self.params["folder_name"] + "_iterations",
+        )
+
+    def data_to_export_iteration(self):
+        """Returns data for iteration exporting.
+
+        Override to customize data to be exported at each nonlinear iteration.
+
+        Returns:
+            Any type compatible with data argument of pp.Exporter().write_vtu().
+
+        """
+        return self.data_to_export()
+
+    def save_data_iteration(self):
+        """Export current solution to vtu files.
+
+        This method is typically called by after_nonlinear_iteration.
+        """
+        # To make sure the nonlinear iteration index does not interfere with the time
+        # part, we multiply the latter by the next power of ten above the maximum number
+        # of nonlinear iterations. Default value set to 10 in accordance with the
+        # default value used in NewtonSolver.
+        n = self.params.get("max_iterations", 10)
+        p = round(np.log10(n))
+        r = 10**p
+        if r <= n:
+            r = 10 ** (p + 1)
+        self.iteration_exporter.write_vtu(
+            self.data_to_export_iteration(),
+            time_dependent=True,
+            time_step=self.nonlinear_solver_statistics.num_iteration
+            + r * self.time_manager.time_index,
+        )
+
+    def after_nonlinear_iteration(self, solution_vector: np.ndarray) -> None:
+        """Integrate iteration export into simulation workflow.
+
+        Order of operations is important, super call distributes the solution to
+        iterate subdictionary.
+
+        """
+        super().after_nonlinear_iteration(solution_vector)  # type: ignore[misc]
+        self.save_data_iteration()
+        self.iteration_exporter.write_pvd()
+
+
+class ResidualExporting:
+    if TYPE_CHECKING:
+        equation_system: pp.EquationSystem
+
+    def data_to_export(self) -> list[DataInput]:
+        """Return data to be exported, including residuals.
+
+        Returns:
+            List containing all (grid, name, values) tuples.
+
+        """
+        data = super().data_to_export()  # type: ignore[misc]
+
+        # Add residuals. Loop over equations in the equation system.
+        for name, operator in self.equation_system.equations.items():
+            residuals = cast(NDArray, self.equation_system.evaluate(operator))
+
+            # Get image_info as dict[GridEntity, int], where
+            # GridEntity = Literal["cells", "faces", "nodes"]
+            image_info = self.equation_system._equation_image_size_info[name]
+            dof_start, dof_end = 0, 0
```

Comment:
Similar to the DeformationExporting, why not pulling the evaluation of operator over [g] inside the for-loop, and getting rid of the explicit dof management?